rust/agents/relayer/src/msg/mod.rs

```diff
@@ -0,0 +1,66 @@
+use std::cmp::Ordering;
+
+use abacus_core::{accumulator::merkle::Proof, CommittedMessage, MultisigSignedCheckpoint};
+
+pub mod gelato_submitter;
+pub mod processor;
+pub mod serial_submitter;
+
+/// processor scans DB for new messages and sends relevant messages
+/// over a channel to a submitter, for delivery.
+///
+/// a submitter uses some strategy to try to delivery those messages
+/// to the target blockchain.
+///
+/// a SubmitMessageOp describes the message that the submitter should
+/// try to submit.
+///
+/// right now there are two strategies: serial and gelato.
+///
+/// in the future it could make sense for there to be more, some ideas are:
+///   - BatchingMessagesSubmitter
+///   - ShardedWalletSubmitter (to get parallelism / nonce)
+///   - SpeculativeSerializedSubmitter (batches with higher optimistic
+///     nonces, recovery behavior)
+///   - FallbackProviderSubmitter (Serialized, but if some RPC provider sucks,
+///   switch everyone to new one)
+
+#[derive(Clone, Debug)]
+pub struct SubmitMessageArgs {
+    pub leaf_index: u32,
+    pub committed_message: CommittedMessage,
+    pub checkpoint: MultisigSignedCheckpoint,
+    pub proof: Proof,
+    pub num_retries: u32,
+}
+
+// The runqueue implementation is a max-heap.  We want the next op to
+// be min over <num_retries, leaf_index>, so the total ordering is
+// the reverse of the natural lexicographic ordering.
```

Comment:
Will probably end up wrapping this in a Schedulable(SubmitMessageArgs) within serial_submitter.rs, and define the comparator over Schedulable. Would probably also move num_retries into Schedulable since it's relevant for scheduling and not information related to the actual message contents that should be delivered. We can then also define execute() over Schedulable and test the thing without having to depend on actually submitting any real ops. Open to other ideas for organizing though, lmk.

Reply:
I agree with what you have said and have placed in a TODO.